### DISCUSSION TITLE
Training Neural ODEs with collocation

### DISCUSSION THREAD

**Discussion by IlyaOrson** [infiniteopt/InfiniteOpt.jl#207] [Q&A] [answered]
I am working on a parameter estimation problem. The parameters are used within a function that takes as input the dynamical variables. I tried to model this function as a parameter function because I would like to include its output also in the constraints and the objective of the problem. Is this possible?

I am currently getting this error: 
```julia
@parameter_function(model, c == custom_function(x, p))`: Cannot have mixed parameter types in a tuple element and can only specify infinite parameters.
```
where `x` is a dynamic variable with support in an infinite parameter `t`, and the parameters `p` are set as independent infinite parameters.

**Comment by IlyaOrson** [comment]
To be more specific here is my concrete example:
```julia
using InfiniteOpt, Ipopt

u0 = [0.0f0, 1.0f0]
tspan = (0,1)
nparams = 2
nstates = 2
nodes_per_element = 2
time_supports = 10
param_supports = 2

function custom_function(x,p)
    reshape(p, 1, nstates) * x
end

optimizer = optimizer_with_attributes(Ipopt.Optimizer, "print_level" => 0)
model = InfiniteModel(optimizer)
method = OrthogonalCollocation(nodes_per_element)

@infinite_parameter(
    model, t in [tspan[1], tspan[2]], num_supports = time_supports, derivative_method = method
)
@infinite_parameter(model, p[1:nparams] in [-1, 1], independent = true, num_supports = param_supports)

@variables(
    model,
    begin
        x[1:2], Infinite(t)  # dynamic variables
    end
)

@parameter_function(model, c == custom_function(x, p))
# custom_function(x, p)  # trace
# @register(model, custom_function(x, p))

# initial conditions
@constraint(model, [i = 1:2], x[i](0) == u0[i])

# variable constraints
@constraint(model, -0.3 <= c(x, p)[1] <= 1.0)
@constraint(model, -0.4 <= x[1])

# dynamic equations
@constraints(
    model,
    begin
        ∂(x[1], t) == (1 - x[2]^2) * x[1] - x[2] + c(x,p)[1]
        ∂(x[2], t) == x[1]
    end
)

@objective(model, Min, integral(x[1]^2 + x[2]^2, t))  # + c(x, p)[1]^2

optimize!(model)
jump_model = optimizer_model(model)
solution_summary(jump_model; verbose=false)
```

**Comment by pulsipher** [comment, marked as answer]
The parameter function syntax just provides a way to implement arbitrary functions that *only* depend on infinite parameters (no variables allowed). Functions of variables should be expressed algebraically if at all possible. In the above case it looks like you can just use an inner product instead `p' * x`.

We can also register functions following our nonlinear syntax, but this should be only done as a last resort (since JuMP does not give 2nd order derivatives for multivariate registered functions). Note that in like manner to JuMP, we do not currently support vector valued inputs for registered functions. Though we plan to support this in the future.

**Comment by IlyaOrson** [comment]
Thank you for the clarification. I was missusing the infinite parameters though, those should have been just classical decision variables without an infinite support for a traditional parameter estimation problem.

My real function is nonlinear and unfortunately vectorized. I tried to follow the [JUMP recommendation](https://jump.dev/JuMP.jl/stable/manual/nlp/#User-defined-functions-with-vector-inputs) and posed it as a multiargument function with splatting instead and now the MWE seems to be working!

```julia
using InfiniteOpt, Ipopt

u0 = [0.0f0, 1.0f0]
tspan = (0,1)
nstates = 2
nparams = 2
nodes_per_element = 2
time_supports = 10

function custom_function(x,p)
    tanh.(reshape(p, 1, nstates) * x)
end

optimizer = optimizer_with_attributes(Ipopt.Optimizer, "print_level" => 0)
model = InfiniteModel(optimizer)
method = OrthogonalCollocation(nodes_per_element)

@infinite_parameter(
    model, t in [tspan[1], tspan[2]], num_supports = time_supports, derivative_method = method
)
# classic fixed-time decision variables
@variable(model, -1 <= p[i = 1:nparams] <= 1)

@variables(
    model,
    begin
        x[1:2], Infinite(t)  # dynamic variables
    end
)

function scalar_fun(z...)
    # @show z typeof(z) z[1:nstates] z[nstates+1:end]
    x = collect(z[1:nstates])
    p = collect(z[nstates+1:end])
    custom_function(x, p)
end

# initial conditions
@constraint(model, [i = 1:2], x[i](0) == u0[i])

# variable constraints
@constraint(model, -0.4 <= x[1])

# dynamic equations
@constraints(
    model,
    begin
        ∂(x[1], t) == (1 - x[2]^2) * x[1] - x[2] + scalar_fun(vcat(x, p)...)[1]
        ∂(x[2], t) == x[1]
    end
)

@objective(model, Min, integral(x[1]^2 + x[2]^2, t))  # + scalar_fun(vcat(x, p)...)[1]^2

optimize!(model)
jump_model = optimizer_model(model)
solution_summary(jump_model; verbose=false)
```
Thanks!

**Comment by IlyaOrson** [comment]
This can be used to train NeuralODEs. It replaces ODE integration and the use of the adjoint method with collocation, training concurrently the NN parameters and the collocation parameters.

This was way more reliable to converge from cold-starts than the adjoint method used in NeuralODEs literature. The trade-off is probably scalability, but it could be a good fit for the smaller NNs often used in SciML.

I've put together a [proof-of-concept](https://github.com/ilyaorson/directcollocationneuralodes) using InfiniteOpt.jl in a self-contained Pluto notebook to demonstrate the idea on an optimal control problem with constraints in the van der Pol system.

https://ilyaorson.github.io/DirectCollocationNeuralODEs/

Thought it could be of interest @pulsipher

**Reply by pulsipher** [reply]
Very cool @IlyaOrson, this reminds me of some recent articles I saw (both of which used Python):
-https://arxiv.org/abs/2502.15642
-https://arxiv.org/abs/2504.04665
It is cool to see a similar approach using InfiniteOpt!

**Reply by IlyaOrson** [reply]
Huh, I should have published this a few years ago with Julia instead 😅 Oh well.

Seemed to me back then that the major advantage was the effective path constraint handling (+ converging at all). This was a major pain point to enforce with the adjoint method and opens up online applications like NMPC.

Cheers for this package btw!

**Reply by pulsipher** [reply]
As a follow up, with #390, we can now import trained NNs into InfiniteOpt to support objects like neural ODEs. An interesting further development would be to import these with their parameters as variables such that they can be trained as well.
